lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+  the abstract guard usually only depends on abstract values and the concrete guard only on
+  concrete values. If we make a case distinction on such an abstract value, the concrete guard
+  cannot to transformed into one that follows that case distinction, so @{method wpc} will fail.
+  The best we can do in this case is give each concrete case a different guard (but still one that
+  does not depend the abstract value). This is fundamental to our corres definition and seems to be
+  sufficient in our manual proofs.
+
+  Extending @{method wpc} so far that it can deal with both situations (uniform treatment of
+  guards/preconditions for e.g. @{term valid} and @{term ccorres}, and non-uniform treatment for
+  things like @{term corres}) would potentially be possible, but would make setup for new constants
+  even more messy than it already is, so we instead re-use the general idea here (in Eisbach instead
+  of ML), and leave the @{method wpc} setup clean for other uses.
+\<close>
+
+section \<open>Helper functions and definitions\<close>
+
+(* The following three definitions are originally by Dan Matichuck from the Eisbach
+   Corres_Method example *)
+
+(* Retrieve a split rule for a target term that is expected to be a case statement. *)
+ML \<open>
+fun get_split_rule ctxt target =
+let
+  val (hdTarget,args) = strip_comb (Envir.eta_contract target)
+  val (constNm, _)  = dest_Const hdTarget
+  val constNm_fds   = (String.fields (fn c => c = #".") constNm)
+
+  val _ = if String.isPrefix "case_" (List.last constNm_fds) then ()
+          else raise TERM ("Not a case statement",[target])
+
+  val typeNm        = (String.concatWith "." o rev o tl o rev) constNm_fds;
+  val split         = Proof_Context.get_thm ctxt (typeNm ^ ".split");
+  val vars = Term.add_vars (Thm.prop_of split) []
+
+  val datatype_name = List.nth (rev constNm_fds,1)
+
+  fun T_is_datatype (Type (nm,_)) = (Long_Name.base_name nm) = (Long_Name.base_name datatype_name)
+    | T_is_datatype _ = false
+
+  val datatype_var =
+    case (find_first (fn ((_,_),T') => (T_is_datatype T')) vars) of
+      SOME (ix,_) => ix
+    | NONE => error ("Couldn't find datatype in thm: " ^ datatype_name)
+
+  val split' = Drule.infer_instantiate ctxt
+    [(datatype_var, Thm.cterm_of ctxt (List.last args))] split
+
+in
+   SOME split' end
+   handle TERM _ => NONE;
+\<close>
+
+(* The above function as an attribute. The term argument is expected to be a case statement. *)
+attribute_setup get_split_rule = \<open>Args.term >>
+  (fn t => Thm.rule_attribute [] (fn context => fn _ =>
+      case (get_split_rule (Context.proof_of context) t) of
+        SOME thm => thm
+      | NONE => Drule.free_dummy_thm))\<close>
+
+(* Apply a split rule to a goal. Example usage:
+
+   apply_split f "\<lambda>f. corres_underlying sr nf nf' r P P' f f'"
+
+   The corres_underlying term provides the context, and f is the term we expect to be
+   a case statement.  *)
```

Comment:
does the f in `f` and f in `\lambda f. ?whatever f` have the same significance relative to the `f` mentioned in "and f is the term we expect ..."?

Reply:
The lambda describes where the `f` is expected inside the bigger term and the `f` itself is the term we expect to be in the form of a case statement.

Reply:
But if we have a bind, then `case ...  >>= return ...` would be `f` and not a case statement? Am I being obtuse here?

Reply:
The case for bind looks like this:

```
apply_split f "λf. corres_underlying sr nf nf' r P P' (f >>= g) f'"
```

so the `f` is still the case statement, but the larger context term has changed to include the bind.

Reply:
Hmm, I find that a bit confusing. It's a lot of `f`.

Reply:
There are only two `f``, one bound, one free. The free one is expected to be `case ..` the bound one says where in the goal that `case ..` term is.

Reply:
(have updated the explanation with a bit more detail)